lib/appsignal/rack/generic_instrumentation.rb

```diff
@@ -1,41 +1,104 @@
 # frozen_string_literal: true
 
 require "rack"
+require "delegate"
 
 module Appsignal
   # @api private
   module Rack
     class GenericInstrumentation
+      # This wrapper is necessary because the request may get initialized in an outer middleware, before
+      # the ActionDispatch request gets created. In that case, the Rack request will take precedence
+      # and the Rails filtered params are not going to be honored, potentially revealing user data.
+      # This wrapper will check whether the actual Rack env contains an ActionDispatch request
+      # and delegate the "filtered_params" call to it when the transaction gets completed. This will
+      # ensure that if there was a Rails request somewhere in the handling flow upstream from this
+      # middleware, its filtered params will end up used for the Appsignal transaction. So barring
+      # a few edge cases even when GenericInstrumentation gets mounted from a Rails app' `config.ru`
+      # there will still be no involuntary disclosure of data that is supposed to get filtered
+      class FilteredParamsWrapper < SimpleDelegator
+        def filtered_params
+          actual_request = __getobj__
+          if actual_request.respond_to?(:filtered_params)
+            actual_request.filtered_params
+          elsif has_rails_filtered_params?(actual_request)
+            # This will delegate to the request itself in case of Rails
+            actual_request.env["action_dispatch.request"].filtered_params
```

Comment:
I hope we can move these kinds of checks to the middleware of library they're for. That way the instrumentation of each library is together with all the other instrumentation for that library and easier to find back.

See also my comment on: https://github.com/appsignal/appsignal-ruby/issues/329#issuecomment-1964111012

Reply:
Yeah this is a side-effect of trying to use the `filtered_params` no-matter-what. If `request` may be re-assigned this is moot (and too clever).